examples/chip-tool/py_matter_chip_tool_adapter/matter_chip_tool_adapter/decoder.py

```diff
@@ -0,0 +1,292 @@
+        elif value is not None and typename.lower() == 'single':
+            value = float('%g' % value)
+
+        return value
+
+
+class OctetStringConverter():
+    def run(self, specs, value, cluster_name: str, typename: str, array: bool):
+        if isinstance(value, dict) and not array:
+            struct = specs.get_struct_by_name(
+                cluster_name, typename) or specs.get_event_by_name(cluster_name, typename)
+            for field in struct.fields:
+                field_name = field.name
+                field_type = field.data_type.name
+                field_array = field.is_list
+                if field_name in value:
+                    value[field_name] = self.run(
+                        specs, value[field_name], cluster_name, field_type, field_array)
+        elif isinstance(value, list) and array:
+            value = [self.run(specs, v, cluster_name, typename, False)
+                     for v in value]
+        elif isinstance(value, str) and (typename.lower() == 'octet_string' or typename.lower() == 'long_octet_string'):
+            if value == '':
+                value = bytes()
+            elif value.startswith('base64:'):
+                value = base64.b64decode(value[7:])
+
+        return value
+
+
+class StructFieldsNameConverter():
+    """
+    Converts fields identifiers to the fields names specified in the cluster definition.
+    """
+
+    def run(self, specs, value, cluster_name: str, typename: str, array: bool):
+        if isinstance(value, dict) and not array:
+            struct = specs.get_struct_by_name(
+                cluster_name, typename) or specs.get_event_by_name(cluster_name, typename)
+            for field in struct.fields:
+                field_code = field.code
+                field_name = field.name
+                field_type = field.data_type.name
+                field_array = field.is_list
+                field_nullable = specs.is_nullable(field)
+                # chip-tool returns the field code as an integer but the test suite expect
+                # field name.
+                # To not confuse the test suite, the field code is replaced by its field name
+                # equivalent and then removed.
+                if str(field_code) in value:
+                    value[field_name] = self.run(
+                        specs,
+                        value[str(field_code)],
+                        cluster_name,
+                        field_type,
+                        field_array
+                    )
+                    del value[str(field_code)]
+                elif field_nullable:
+                    # If the fields mapping contains additional values that are not contains
+                    # in the response, we need to check if those are Nullables.
+                    # This is because nullable values are not part of the TLV, even if they
+                    # are not Optionals. That may confuse the test suite that expects those
```

Comment:
Ah right. That was truly a bug until your changes at https://github.com/project-chip/connectedhomeip/pull/25092
I have removed the code in python that was assuming that, it will let us catch those types of things...